### ISSUE TITLE
2.3.15 R4 - Not running Ubuntu 19.1

### ISSUE THREAD

**Issue by foss365** [NatronGitHub/Natron#410]
### Problem

2.3.15 R4 will not run.

2.3.15 RC6 has to be tested

[EDIT] We found the reasons, now let's solve the various issues:

~~1. upload strings32.tgz and strings64.tgz to the two locations https://github.com/NatronGitHub/NatronGitHub.github.io/tree/master/files/bin and https://sourceforge.net/projects/natron/files/Third_Party/Binaries/~~
~~2. understand and fix why python27.zip and PySide are not installed~~
~~3. check if the Natron launch script needs to be updated for recent libstdc++ versions, and that it does the same mumbo-jumbo with libgcc_s.so (which has the same kind of internal-versioning issues).~~

**Expected behavior:** [What you expected to happen]

Application to start up

**Actual behavior:** [What actually happened]

 Won't start has the following errors in the CMD window

/home/tom/Natron-snapshot/bin/strings: error while loading shared libraries: libbfd-2.20.51.0.2-5.48.el6.so: cannot open shared object file: No such file or directory
Natron Version 2.3.15
Copyright (C) 2013-2018 INRIA and Alexandre Gauthier-Foichat
>>>Use the --help or -h option to print usage.<<<
"/home/tom/Natron-snapshot/bin/../lib/python27.zip" does not exist, not added to PYTHONPATH
Py_SetPythonHome("/home/tom/Natron-snapshot/bin/..")
PYTHONPATH set to /home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../Plugins
Py_SetProgramName("/home/tom/Natron-snapshot/bin/Natron")
Py_Initialize()
ImportError: No module named site

 ### Versions
2.3.15

* **OS version:** 
Linux Ubuntu Disco 19.1

**Comment by rodlie** [comment]
Will be fixed.

'strings' needs to be built on host (enable-static), I forgot that CentOS uses a "custom" bfd lib.

**Comment by rodlie** [comment]
@foss365 : Please download this file and replace NATRON_FOLDER/bin/strings
[natron-strings-linux64.zip](https://github.com/NatronGitHub/Natron/files/3436143/natron-strings-linux64.zip)

That should fix the issue.

**Comment by devernay** [comment]
@rodlie why exactly does the script have to run `strings`?
please upload strings32.tgz and strigs64.tgz to https://github.com/NatronGitHub/NatronGitHub.github.io/tree/master/files/bin and to https://sourceforge.net/projects/natron/files/Third_Party/Binaries/ as well as any other "pre-built binary" that is used during the build process. I loast most of the files when the natron.fr server went down

**Comment by rodlie** [comment]
strings are used to get GLIBCXX in Natron-Linux.sh.

'strings' are now built in the SDK, no need for strings*.tgz. I have updated binaries for qtifw for Windows, will upload when I have tested the final installer.

**Comment by devernay** [comment]
I don't think that's a good way to proceed. No need to play with libstdc++, we ship it with the binary

**Comment by rodlie** [comment]
The best solution would be to use DTS on CentOS, that way we don't have to care about libstdc++ etc. On some system the supplied libstdc++ will create problems (we should never override system libstdc++).

But, not my call. You could of course just remove everything related to libstdc++ checking in Natron-Linux.sh, it should work on the majority of system.

**Comment by devernay** [comment]
libstdc++ is *not* part of the system, only libc is. libstdc++ is a userland library, and we can ship our version of it, which is known to work with *our* binaries.
If you do that for libstdc++, then why not for *all* other dynamic libraries, which are userland libraries just like libstdc++?

**Comment by rodlie** [comment]
But not known to work with the system we are running on. We force a library that might conflict with system software (like drivers) when we load Natron.

**Comment by devernay** [comment]
Drivers should never use dynamic userland libraries. This is how Unix is designed. Drivers have to use system libraries only. If there is a specific issue with a specific driver (AMD?), then it's ok to provide an optional fix that works out this specific issue , but there is no generic solution if badly designed drivers link to userland dynamic libraries that date supposed to be provided by the "system". libstdc++ is not the only userland dynamic library we provide.

**Comment by rodlie** [comment]
Mesa supplies DRI drivers on Linux, they depend on libstdc++.

**Comment by foss365** [comment]
That got rid of the strings error but I still have the following errors and it won't start.

Natron Version 2.3.15
Copyright (C) 2013-2018 INRIA and Alexandre Gauthier-Foichat
>>>Use the --help or -h option to print usage.<<<
"/home/tom/Natron-snapshot/bin/../lib/python27.zip" does not exist, not added to PYTHONPATH
Py_SetPythonHome("/home/tom/Natron-snapshot/bin/..")
PYTHONPATH set to /home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../Plugins
Py_SetProgramName("/home/tom/Natron-snapshot/bin/Natron")
Py_Initialize()
ImportError: No module named site

**Comment by foss365** [comment]
I'm in no way complaining just want to help in any way I can.

**Comment by devernay** [comment]
> "/home/tom/Natron-snapshot/bin/../lib/python27.zip" does not exist

That's the issue, I'll look into that this week-end. You can also copy that file (python27.zip) from a previous Natron installation

**Comment by rodlie** [comment]
the installer is missing the zip, the no-installer has the zip.

**Comment by foss365** [comment]
Zipping up the Python folder in lib and renaming it python27.zip got me less errors but still no go.

Natron Version 2.3.15
Copyright (C) 2013-2018 INRIA and Alexandre Gauthier-Foichat
>>>Use the --help or -h option to print usage.<<<
Py_SetPythonHome("/home/tom/Natron-snapshot/bin/..")
PYTHONPATH set to /home/tom/Natron-snapshot/bin/../lib/python27.zip:/home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../Plugins
Py_SetProgramName("/home/tom/Natron-snapshot/bin/Natron")
Py_Initialize()
ImportError: No module named site

**Comment by rodlie** [comment]
PySide is also missing from the installer (you can copy it from the no-installer).

**Comment by foss365** [comment]
Okay copied the pythom27.zip from the no install and got farther. Here are the errors now.

Natron Version 2.3.15
Copyright (C) 2013-2018 INRIA and Alexandre Gauthier-Foichat
>>>Use the --help or -h option to print usage.<<<
Py_SetPythonHome("/home/tom/Natron-snapshot/bin/..")
PYTHONPATH set to /home/tom/Natron-snapshot/bin/../lib/python27.zip:/home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../Plugins
Py_SetProgramName("/home/tom/Natron-snapshot/bin/Natron")
Py_Initialize()
PATH is /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/local/cuda/bin:/usr/local/cuda/bin
PYTHONPATH is /home/tom/Natron-snapshot/bin/../lib/python27.zip:/home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../Plugins
PYTHONHOME is (null)
Py_DebugFlag is 0
Py_VerboseFlag is 0
Py_InteractiveFlag is 0
Py_InspectFlag is 0
Py_OptimizeFlag is 0
Py_NoSiteFlag is 0
Py_BytesWarningFlag is 0
Py_UseClassExceptionsFlag is 1
Py_FrozenFlag is 0
Py_TabcheckFlag is 0
Py_UnicodeFlag is 0
Py_IgnoreEnvironmentFlag is 0
Py_DivisionWarningFlag is 0
Py_DontWriteBytecodeFlag is 0
Py_NoUserSiteDirectory is 1
Py_GetProgramName is /home/tom/Natron-snapshot/bin/Natron
Py_GetPrefix is /home/tom/Natron-snapshot/bin/..
Py_GetExecPrefix is /home/tom/Natron-snapshot/bin/..
Py_GetProgramFullPath is /home/tom/Natron-snapshot/bin/Natron
Py_GetPath is /home/tom/Natron-snapshot/bin/../lib/python27.zip:/home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../Plugins:/home/tom/Natron-snapshot/bin/../lib/python27.zip:/home/tom/Natron-snapshot/bin/../lib/python2.7:/home/tom/Natron-snapshot/bin/../lib/python2.7/plat-linux2:/home/tom/Natron-snapshot/bin/../lib/python2.7/lib-tk:/home/tom/Natron-snapshot/bin/../lib/python2.7/lib-old:/home/tom/Natron-snapshot/bin/../lib/python2.7/lib-dynload
Py_GetPythonHome is /home/tom/Natron-snapshot/bin/..
Python library is in /home/tom/Natron-snapshot/lib/python2.7/site-packages
Error while loading python module NatronEngine: While executing script:
import sys
from math import *
import NatronEnginePython error:
Python exception: could not import module 'PySide.QtCore'
Traceback (most recent call last):
  File "<string>", line 3, in <module>
ImportError: could not import module 'PySide.QtCore'


Segmentation fault (core dumped)

**Comment by devernay** [comment]
Maybe I'll remove the installer for now.
Also that segfault is not a good thing, I'll have to fix that in the code (it should just throw an exception).

**Comment by foss365** [comment]
Yes! The no Install version with the updated natron-strings-linux64.zip above worked fine.

**Comment by foss365** [comment]
![Screenshot_2019-07-26_16-42-20](https://user-images.githubusercontent.com/16073383/61983126-bcf2f280-afc4-11e9-8993-13fc411a26d9.png)

**Comment by devernay** [comment]
Not closing this issue until those are done:
1. upload strings32.tgz and strings64.tgz to the two locations  https://github.com/NatronGitHub/NatronGitHub.github.io/tree/master/files/bin and  https://sourceforge.net/projects/natron/files/Third_Party/Binaries/
2. understand and fix why python27.zip is not installed
3. check if the Natron launch script needs to be updated for recent libstdc++ versions, and that it does the same mumbo-jumbo with libgcc_s.so (which has the same kind of internal-versioning issues) [1].

@rodlie can you please take this from here?

[1] It seems that VMWare implements a similar thing, maybe we could check how the VMWare launch script solves that. Also Steam seems to have had this issue - how was it solved in the end? https://github.com/ValveSoftware/steam-runtime/issues/13

**Comment by rodlie** [comment]
Will fix the Linux issues, I'm currently 99% done with Windows, just need to fix some missing dll's and run the full unit tests.

**Comment by devernay** [comment]
Be careful that the unit tests report may be wrong, especially if the test does not render any image (that happened when Shadertoy was not built with mesa support).
Look at the full test log.
Baymax is the only one that fails on Linux, and I didn't have time to dive deep into that.

**Comment by devernay** [comment]
Actually, looking back at the linux launch script, we haven't been using "strings" since Oct 24 2017, i.e. version 2.3.3.
I had found a much better solution at that time, which was to use patchelf and never set LD_LIBRARY_PATH.
And nobody has been complaining since then, so I'm removing that part of the launch script, and remove the installation of "strings" alltogether.
Only [2] is left to fix.

**Comment by rodlie** [comment]
Ok, I just started a Linux build and will check the missing python.zip.

The Windows builds seems to be fine, but the unit tests script seems to fail even before running any tests (currently waiting on a new build), will find the issue.

**Comment by devernay** [comment]
note that in the SDK PySide was previously installed in SDKDIR, and it is now installed in SDKDIR/qt4. This may have broken PySide for the installer version (I fixed the location at two places, but maybe it was hardcoded at other places):  https://github.com/NatronGitHub/Natron/commit/d69fc37b5b8641c330862a8d30dbf47005c39280

**Comment by rodlie** [comment]
Something is reverting my modifications (or local commits) when I run the build scripts, so I can't get a proper package build. This diff should work:
```
diff --git a/tools/jenkins/build-Linux-installer.sh b/tools/jenkins/build-Linux-installer.sh
index ad8593c..5edfd82 100755
--- a/tools/jenkins/build-Linux-installer.sh
+++ b/tools/jenkins/build-Linux-installer.sh
@@ -552,6 +552,9 @@ fi
 
 # Copy Python distrib to installer package
 cp -r "$PYDIR" "$LIBS_PACKAGE_PATH/data/lib/"
+cp "${TMP_PORTABLE_DIR}"/lib/python*.zip "${LIBS_PACKAGE_PATH}/data/lib/"
+mkdir -p "${LIBS_PACKAGE_PATH}/data/Plugins"
+cp -a "${TMP_PORTABLE_DIR}/Plugins/PySide" "${LIBS_PACKAGE_PATH}/data/Plugins/"
 
 # Fix RPATH (we don't want to link against system libraries when deployed)
 for location in "$LIBS_PACKAGE_PATH/data" "${TMP_PORTABLE_DIR}"; do

```

**Comment by devernay** [comment]
"something" is the build script itself, which updates scripts from github
Thank you, I'll apply that, but there are also a lot of stuff to fix: all site-packages from the SDK are copied. We should whitelist python packages instead of blacklisting them. I'm on it

**Comment by devernay** [comment]
done. please test

**Comment by rodlie** [comment]
```
+ for pydir in '"${SDK_HOME}/lib/python${PYVER}"' '"${SDK_HOME}/qt${QT_VERSION_MAJOR}/lib/python${PYVER}"'
+ cd /opt/Natron-sdk/lib/python2.7
+ tar cf - . --exclude site-packages
+ cd /home/olear/Natron-snapshot-workspace/tmp/tmp_deploy/Natron-RB-2.3-201907281909--64-no-installer/lib/python2.7
+ tar xf -
+ cd /opt/Natron-sdk/lib/python2.7/site-packages
+ tar cf - easy_install.py pip PySide README shiboken.so
+ cd /home/olear/Natron-snapshot-workspace/tmp/tmp_deploy/Natron-RB-2.3-201907281909--64-no-installer/lib/python2.7/site-packages
+ tar xf -
+ true
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
build-Linux-installer.sh failed with status 1
```

**Comment by devernay** [comment]
should work now (cannot test - I'm rebuilding the sdk)

**Comment by rodlie** [comment]
will test

**Comment by devernay** [comment]
@foss365 Thomas, can you please test with the latest binary installer?

**Comment by foss365** [comment]
This installed fine on Ubuntu 19.x I did already find an issue with moving text around the screen but I'll open another issue for that. You can close this issue, thanks for all your work!
